File: erebus/api/sessions.py
"""Session management for Erebus WebUI.

Provides persistent session storage (JSON files) with in-memory LRU cache,
session CRUD operations, and compact session listing.
"""

from __future__ import annotations

import json
import time
import uuid
from collections import OrderedDict
from dataclasses import asdict, dataclass, field
from pathlib import Path
from threading import Lock
from typing import Any, Optional


@dataclass
class Session:
    """A single chat session."""

    session_id: str
    title: str
    model: str
    messages: list[dict[str, Any]] = field(default_factory=list)
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    pinned: bool = False
    archived: bool = False
    tool_calls: list[dict[str, Any]] = field(default_factory=list)

    def compact(self) -> dict[str, Any]:
        """Return metadata-only dict (no messages) for list views."""
        d = asdict(self)
        d.pop("messages", None)
        d.pop("tool_calls", None)
        d["message_count"] = len(self.messages)
        return d


_MAX_CACHE = 100
_sessions: OrderedDict[str, Session] = OrderedDict()
_lock = Lock()


def _sessions_dir(data_dir: Path) -> Path:
    d = data_dir / "sessions"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _session_path(data_dir: Path, sid: str) -> Path:
    return _sessions_dir(data_dir) / f"{sid}.json"


def save_session(data_dir: Path, session: Session) -> None:
    """Persist session to disk and update cache."""
    session.updated_at = time.time()
    path = _session_path(data_dir, session.session_id)
    path.write_text(json.dumps(asdict(session), default=str), encoding="utf-8")
    with _lock:
        _sessions[session.session_id] = session
        _sessions.move_to_end(session.session_id)
        while len(_sessions) > _MAX_CACHE:
            _sessions.popitem(last=False)


def load_session(data_dir: Path, sid: str) -> Optional[Session]:
    """Load session from cache or disk."""
    with _lock:
        if sid in _sessions:
            _sessions.move_to_end(sid)
            return _sessions[sid]

    path = _session_path(data_dir, sid)
    if not path.exists():
        return None
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
        if not isinstance(raw, dict):
            return None
        # Validate required fields exist
        if "session_id" not in raw or "title" not in raw or "model" not in raw:
            return None
        filtered = {k: v for k, v in raw.items() if k in Session.__dataclass_fields__}
        session = Session(**filtered)
        with _lock:
            _sessions[sid] = session
            while len(_sessions) > _MAX_CACHE:
                _sessions.popitem(last=False)
        return session
    except (json.JSONDecodeError, TypeError, ValueError):
        return None


def new_session(data_dir: Path, model: str, title: str = "New Chat") -> Session:
    """Create and persist a new session."""
    sid = uuid.uuid4().hex[:12]
    session = Session(session_id=sid, title=title, model=model)
    save_session(data_dir, session)
    return session


def delete_session(data_dir: Path, sid: str) -> bool:
    """Delete a session from disk and cache."""
    with _lock:
        _sessions.pop(sid, None)
    path = _session_path(data_dir, sid)
    if path.exists():
        path.unlink()
        return True
    return False


def all_sessions(data_dir: Path) -> list[dict[str, Any]]:
    """Return compact metadata for all sessions, sorted by updated_at desc."""
    sessions_dir = _sessions_dir(data_dir)
    result = []
    for path in sessions_dir.glob("*.json"):
        sid = path.stem
        session = load_session(data_dir, sid)
        if session:
            result.append(session.compact())
    result.sort(key=lambda s: s["updated_at"], reverse=True)
    return result


def rename_session(data_dir: Path, sid: str, title: str) -> Optional[Session]:
    """Rename a session."""
    session = load_session(data_dir, sid)
    if session is None:
        return None
    session.title = title
    save_session(data_dir, session)
    return session

File: erebuslite/internal/sessions/sessions.go
// Package sessions provides persistent chat session storage using JSON files.
package sessions

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

// Session represents a single chat session.
type Session struct {
	SessionID string           `json:"session_id"`
	Title     string           `json:"title"`
	Model     string           `json:"model"`
	Messages  []Message        `json:"messages"`
	CreatedAt float64          `json:"created_at"`
	UpdatedAt float64          `json:"updated_at"`
	Pinned    bool             `json:"pinned"`
	Archived  bool             `json:"archived"`
	ToolCalls []map[string]any `json:"tool_calls"`
}

// Message represents a single message in a session.
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Compact returns metadata-only dict (no messages) for list views.
type SessionCompact struct {
	SessionID    string  `json:"session_id"`
	Title        string  `json:"title"`
	Model        string  `json:"model"`
	MessageCount int     `json:"message_count"`
	CreatedAt    float64 `json:"created_at"`
	UpdatedAt    float64 `json:"updated_at"`
	Pinned       bool    `json:"pinned"`
	Archived     bool    `json:"archived"`
}

// Compact returns a compact representation of the session without messages.
func (s *Session) Compact() SessionCompact {
	return SessionCompact{
		SessionID:    s.SessionID,
		Title:        s.Title,
		Model:        s.Model,
		MessageCount: len(s.Messages),
		CreatedAt:    s.CreatedAt,
		UpdatedAt:    s.UpdatedAt,
		Pinned:       s.Pinned,
		Archived:     s.Archived,
	}
}

// Store manages session persistence and caching.
type Store struct {
	dataDir string
	mu      sync.RWMutex
	cache   map[string]*Session
}

// NewStore creates a new session store.
func NewStore(dataDir string) *Store {
	dir := filepath.Join(dataDir, "sessions")
	_ = os.MkdirAll(dir, 0o755)
	return &Store{
		dataDir: dataDir,
		cache:   make(map[string]*Session),
	}
}

func (s *Store) sessionsDir() string {
	return filepath.Join(s.dataDir, "sessions")
}

func (s *Store) sessionPath(sid string) string {
	return filepath.Join(s.sessionsDir(), sid+".json")
}

// Save persists a session to disk and updates the cache.
func (s *Store) Save(session *Session) error {
	session.UpdatedAt = float64(time.Now().Unix())
	data, err := json.Marshal(session)
	if err != nil {
		return err
	}
	if err := os.WriteFile(s.sessionPath(session.SessionID), data, 0o644); err != nil {
		return err
	}
	s.mu.Lock()
	s.cache[session.SessionID] = session
	s.mu.Unlock()
	return nil
}

// Load retrieves a session from cache or disk.
func (s *Store) Load(sid string) (*Session, error) {
	s.mu.RLock()
	if sess, ok := s.cache[sid]; ok {
		s.mu.RUnlock()
		return sess, nil
	}
	s.mu.RUnlock()

	data, err := os.ReadFile(s.sessionPath(sid))
	if err != nil {
		return nil, err
	}
	var session Session
	if err := json.Unmarshal(data, &session); err != nil {
		return nil, err
	}
	s.mu.Lock()
	s.cache[sid] = &session
	s.mu.Unlock()
	return &session, nil
}

// New creates and persists a new session.
func (s *Store) New(model, title string) (*Session, error) {
	now := float64(time.Now().Unix())
	sid := fmt.Sprintf("%x", time.Now().UnixNano())[:12]
	session := &Session{
		SessionID: sid,
		Title:     title,
		Model:     model,
		Messages:  []Message{},
		CreatedAt: now,
		UpdatedAt: now,
		ToolCalls: []map[string]any{},
	}
	if err := s.Save(session); err != nil {
		return nil, err
	}
	return session, nil
}

// Delete removes a session from disk and cache.
func (s *Store) Delete(sid string) bool {
	s.mu.Lock()
	delete(s.cache, sid)
	s.mu.Unlock()
	path := s.sessionPath(sid)
	if err := os.Remove(path); err != nil {
		return false
	}
	return true
}

// Rename changes the title of a session.
func (s *Store) Rename(sid, title string) (*Session, error) {
	session, err := s.Load(sid)
	if err != nil {
		return nil, err
	}
	session.Title = title
	if err := s.Save(session); err != nil {
		return nil, err
	}
	return session, nil
}

// All returns compact metadata for all sessions, sorted by updated_at desc.
func (s *Store) All() []SessionCompact {
	dir := s.sessionsDir()
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var result []SessionCompact
	for _, entry := range entries {
		if filepath.Ext(entry.Name()) != ".json" {
			continue
		}
		sid := entry.Name()[:len(entry.Name())-5] // strip .json
		session, err := s.Load(sid)
		if err != nil {
			continue
		}
		result = append(result, session.Compact())
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].UpdatedAt > result[j].UpdatedAt
	})
	return result
}
